src/System.Globalization.Calendars/tests/CalendarHelpers.cs

```diff
@@ -2,177 +2,411 @@
+            new ChineseLunisolarCalendar(),
+            new KoreanLunisolarCalendar(),
+            new PersianCalendar()
+        };
+
         private static int MinEra(Calendar calendar) => calendar.GetEra(calendar.MinSupportedDateTime);
+        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
 
-        private static int YearOffset(Calendar calendar) => calendar.GetYear(calendar.MinSupportedDateTime) - 1;
+        private static int MaxCalendarYearInEra(Calendar calendar, int era)
+        {
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[0] || era == 0)
+            {
+                return calendar.GetYear(calendar.MaxSupportedDateTime);
+            }
+            return calendar.GetYear(calendar.ToDateTime(1, 1, 1, 0, 0, 0, 0, era + 1).AddDays(-1)) + 1;
+        }
 
-        public static void AddYears_Invalid(Calendar calendar)
+        private static int MaxGregorianYearInEra(Calendar calendar, int era)
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MaxSupportedDateTime, 1));
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MinSupportedDateTime, -1));
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[0] || era == 0)
+            {
+                return calendar.MaxSupportedDateTime.Year;
+            }
+            return MinGregorianYearInEra(calendar, era + 1);
         }
 
-        public static void AddMonths_Invalid(Calendar calendar)
+        private static int MinGregorianYearInEra(Calendar calendar, int era)
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddMonths(calendar.MaxSupportedDateTime, 1));
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddMonths(calendar.MinSupportedDateTime, -1));
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[eras.Length - 1] || era == 0)
+            {
+                return calendar.MinSupportedDateTime.Year;
+            }
+            return calendar.ToDateTime(1, 1, 1, 0, 0, 0, 0, era).Year;
+        }
 
-            Assert.Throws<ArgumentOutOfRangeException>("months", () => calendar.AddMonths(DateTime.Now, -120001));
-            Assert.Throws<ArgumentOutOfRangeException>("months", () => calendar.AddMonths(DateTime.Now, 120001));
+        private static int MinCalendarYearInEra(Calendar calendar, int era)
+        {
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[eras.Length - 1] || era == 0)
+            {
+                return calendar.GetYear(calendar.MinSupportedDateTime);
+            }
+            return 1;
```

Comment:
This assumption currently is correct for all implementations, but I could see a situation where this could need fixing to be more dynamic